### PR TITLE
CMS: Implement strong params

### PR DESCRIPTION
**Pull request by jlledom**
**What this PR does / why we need it**:

Remove all references to protected attributes from CMS models, and replace them by whitelists of strong params in controllers. 

Rails deprecated protected attributes in version 4, and removed them in version 5, we can use then only because we are using the [protected_attributes_continued](https://github.com/westonganger/protected_attributes_continued) gem.

Since protected attributes were coded at the models, that affected both UI and API. So I had to add the whitelist for controllers in both places. In the particular case of templates controllers, they were implemented in a different approach:

- For the API, there's only one route and one controller to edit all templates: `admin/api/cms/templates_controller.rb`, so I had to add a [dictionary](https://github.com/3scale/porta/pull/3228/files#diff-52b1aa4f8bd23b801e7b72acac67c6aae3e44ae1ee45e40d7da4dcaa3177fad8R5-R12) with the permitted parameters for each type. 
- For the UI, each type (Partial, Page, Layout, etc...) has it's own route and controller, so I took advantage of the class inheritance and implemented the different whitelists by [overriding](https://github.com/3scale/porta/pull/3228/files#diff-e6659bec919c8282861b77d4f7eb03206855afcc14b9a4a53491fbd1bebd8ef0R100-R102) the `allowed_params` method.

On the other hand, after refactoring the API templates controller to allow multiple whitelists, I found a bug in the templates service, that allowed to write on the `type` column. I had to make some changes in the service to prevent it, which consists on passing both permitted and unpermitted params to the service and use only the permitted ones to actually save changes on the DB.

**Which issue(s) this PR fixes** 

[THREESCALE-9079](https://issues.redhat.com/browse/THREESCALE-9079)

**Verification steps** 

Everything should work as before

### CODE REVIEW

File: app/controllers/admin/api/cms/files_controller.rb [hunk 1]
```diff
@@ -31,7 +34,7 @@ class Admin::Api::CMS::FilesController < Admin::Api::CMS::BaseController
   ##~ op.parameters.add @parameter_access_token
   def index
     files = (if params[:section_id]
-      current_account.sections.find_by_id_or_system_name!(params[:section_id]).files
+      current_account.sections.find_by!(id_or_system_name: params[:section_id]).files
```
Comment (jlledom) [inline app/controllers/admin/api/cms/files_controller.rb:37]: This is a Rubocop fix

File: app/controllers/admin/api/cms/templates_controller.rb [hunk 2]
```diff
@@ -48,7 +55,7 @@ def index
   ##~ op.parameters.add :name => "liquid_enabled", :description => "liquid processing of the template content on/off", :paramType => "query", :type => "boolean"
   ##~ op.parameters.add :name => "handler", :paramType => "query", :description => "text will be processed by the handler before rendering", :required => false, :allowableValues => { :valueType => "LIST", :values => ["textile", "markdown"]  }
   def create
-    template = Admin::Api::CMS::TemplateService::Create.call(current_account, cms_template_params)
+    template = Admin::Api::CMS::TemplateService::Create.call(current_account, params, cms_template_params)
```
Comment (jlledom) [inline app/controllers/admin/api/cms/templates_controller.rb:58]: This change comes from https://github.com/3scale/porta/pull/3228/commits/5eba7a04088e05a640bdb90bfcd148f93ff68367. The service was updated to prevent writing the  `type` column.

File: app/controllers/admin/api/cms/sections_controller.rb [hunk 3]
```diff
@@ -2,7 +2,8 @@ class Admin::Api::CMS::SectionsController < Admin::Api::CMS::BaseController
   ##~ sapi = source2swagger.namespace("CMS API")
   ##~ @parameter_section_id = { :name => "id", :description => "ID of the section", :dataType => "int", :required => true, :paramType => "path" }
 
-  wrap_parameters :section, include: [:title, :public, :parent_id, :partial_path]
+  ALLOWED_PARAMS = %i[parent_id title system_name public partial_path].freeze
```
Comment (mayorova) [inline app/controllers/admin/api/cms/sections_controller.rb:5]: Just curious - why the order is like this? 
I mean, it's different compared to the original list in `wrap_parameters`.

Maybe it makes sense to order them alphabetically?

File: app/controllers/provider/admin/cms/builtin_legal_terms_controller.rb [hunk 4]
```diff
@@ -58,6 +58,6 @@ def system_name_param
   end
 
   def permitted_params
-    params.permit(cms_template: %i[system_name draft])
+    params.permit(cms_template: %i[system_name draft published])
```
Comment (mayorova) [inline app/controllers/provider/admin/cms/builtin_legal_terms_controller.rb:61]: Is adding `published` needed?

The controller always received the `draft`, its content is later copied to the `published` field on model save.

File: app/services/admin/api/cms/template_service.rb [hunk 5]
```diff
@@ -8,36 +8,27 @@ class UnknownTemplateTypeError < TemplateServiceError; end
...
   class UnknownLayoutError < TemplateServiceError; end
 
-
-
-  def initialize(current_account, params)
-    @params = params
+  def initialize(current_account, params, permitted_params)
+    @permitted_params = permitted_params
     @current_account = current_account
+    @type = params.delete('type')&.to_sym
```
Comment (mayorova) [inline app/services/admin/api/cms/template_service.rb:14]: Why is this moved to the superclass?
Either `type` or `params` don't seem to be used in the Update service...

File: app/controllers/provider/admin/cms/builtin_legal_terms_controller.rb (hunk 4 above)
Comment (jlledom) [reply app/controllers/provider/admin/cms/builtin_legal_terms_controller.rb:61]: Same as above, I pasted from the [model](https://github.com/3scale/porta/pull/3228/files#diff-3f006b9242fc9ffaa847d7677bbdea9d1b86adef6f462be03e80131417e4a46a). But yes, for this particular controller probably it's not needed, I'll check.

File: features/step_definitions/cms/section_steps.rb [hunk 6]
```diff
@@ -21,11 +21,11 @@
 end
 
 def add_section_to_provider(provider, protection, name, path = nil)
-  root = provider.sections.root || FactoryBot.create(:root_cms_section, :account => provider)
+  root = provider.sections.root || FactoryBot.create(:root_cms_section, provider: provider)
 
   options = {:title => name,
     :system_name => name,
-    :account => provider,
+    :provider => provider,
```
Comment (mayorova) [inline features/step_definitions/cms/section_steps.rb:28]: Maybe while on it, change the syntax of the whole hash?

File: app/services/admin/api/cms/template_service.rb (hunk 5 above)
Comment (jlledom) [reply app/services/admin/api/cms/template_service.rb:14]: You're right, I'll double check this. The reason is I don't want the `type` param to be sent to the `update` method because that writes the `type` column in the DB. But after separating `permitted_params` from `params` this is not needed.

File: features/step_definitions/cms/section_steps.rb (hunk 6 above)
Comment (jlledom) [reply features/step_definitions/cms/section_steps.rb:28]: I thought it was not directly related and could make the review harder

File: app/controllers/provider/admin/cms/sections_controller.rb [hunk 7]
```diff
@@ -64,11 +64,7 @@ def destroy
   protected
 
   def section_params
-    params[:cms_section].dup.tap do |params|
-      params[:parent] = if id = params[:parent_id]
-                          current_account.sections.find(id)
-                        end
-    end
+    params.require(:cms_section).permit(:parent_id, :title, :public, :partial_path)
```
Comment (mayorova) [inline app/controllers/provider/admin/cms/sections_controller.rb:67]: This change is a bit dangerous, because it allows setting a section that belongs to a different provider.
1. It's only possible when hacking the HTML
2. But it may lead to WTF moments, and in general crossing the tenant borders is not good.

File: app/controllers/provider/admin/cms/files_controller.rb [hunk 8]
```diff
@@ -48,11 +48,7 @@ def destroy
   private
 
   def file_params
-    params[:cms_file].dup.tap do |params|
-      if section = params[:section_id]
-        params[:section] = current_account.sections.find(section)
-      end
-    end
+    params.require(:cms_file).permit(:path, :attachment, :downloadable, :tag_list, :section_id)
```
Comment (mayorova) [inline app/controllers/provider/admin/cms/files_controller.rb:51]: Same as https://github.com/3scale/porta/pull/3228/files#r1124388383

Can set the section that is from a different tenant.

File: app/controllers/provider/admin/cms/pages_controller.rb [hunk 9]
```diff
@@ -6,18 +8,7 @@ def templates
...
-
-      section_id = params.delete(:section_id).presence
-      layout_id  = params.delete(:layout_id).presence
-
-      params[:section] = current_account.sections.find(section_id) if section_id
-      # We need to handle the case when the layout_id is purposely nil
-      params[:layout] = layout_id && current_account.layouts.find(layout_id)
-    end
+  def allowed_params
+    %i[title section_id layout_id path content_type tag_list system_name liquid_enabled handler draft].freeze
```
Comment (mayorova) [inline app/controllers/provider/admin/cms/pages_controller.rb:12]: Same as https://github.com/3scale/porta/pull/3228/files#r1124388383

File: app/controllers/admin/api/cms/templates_controller.rb [hunk 10]
```diff
@@ -2,12 +2,19 @@ class Admin::Api::CMS::TemplatesController < Admin::Api::CMS::BaseController
   ##~ sapi = source2swagger.namespace("CMS API")
   ##~ @parameter_template_id = { :name => "id", :description => "ID of the template", :dataType => "int", :required => true, :paramType => "path" }
 
-  ALLOWED_PARAMS = %i[type system_name title path draft liquid_enabled handler content_type section_id section_name layout_id layout_name].freeze
+  AVAILABLE_PARAMS = %i[system_name title path draft liquid_enabled handler content_type section_id layout_id].freeze
```
Comment (mayorova) [inline app/controllers/admin/api/cms/templates_controller.rb:5]: With this change, the possibility to update the templates' layout and section using the system names (`layout_name`, `section_name`) is removed. Is that deliberate?

File: app/controllers/provider/admin/cms/sections_controller.rb (hunk 7 above)
Comment (jlledom) [reply app/controllers/provider/admin/cms/sections_controller.rb:67]: You're right, and now I wonder if this could happen also from the API endpoint. I'll take a look. Thanks

Comment (mayorova) [reply app/controllers/provider/admin/cms/sections_controller.rb:67]: No, I think in the API the logic is correct.

File: app/controllers/admin/api/cms/files_controller.rb [hunk 11]
```diff
@@ -11,9 +11,12 @@ class Admin::Api::CMS::FilesController < Admin::Api::CMS::BaseController
 
   MAX_PER_PAGE = 100
   DEFAULT_PER_PAGE = 20
+  ALLOWED_PARAMS = %i[section_id path attachment downloadable tag_list].freeze
```
Comment (thalesmiguel) [inline app/controllers/admin/api/cms/files_controller.rb:14]: I tried to test this by updating a File's tag_list but `acts_as_taggable` something something doesn't really want me to.
If we're expecting an array of values for `tag_list`, like
```json
{
  "file": [
    "best_tag_ever",
    "not_the_best_tag_ever"
  ]
}
```
I think we need to allow it with `tag_list: []`.
This came out of my not reliable memory so it could be wrong.

File: app/controllers/admin/api/cms/files_controller.rb [hunk 12]
```diff
@@ -11,9 +11,12 @@ class Admin::Api::CMS::FilesController < Admin::Api::CMS::BaseController
...
 
   before_action :find_file, only: [:show, :edit, :update, :destroy]
 
+  wrap_parameters :file, include: ALLOWED_PARAMS
```
Comment (thalesmiguel) [inline app/controllers/admin/api/cms/files_controller.rb:18]: Just to be sure: This is in here so whoever calls this endpoint doesn't have to wrap the body on a `"file":` key, correct? Also, if the body comes wrapped, this will be ignored, right?

File: app/models/cms/section.rb [hunk 13]
```diff
@@ -204,6 +204,12 @@ def not_own_child
     end
   end
 
+  def parent_same_provider
+    return if parent&.provider == provider
```
Comment (thalesmiguel) [inline app/models/cms/section.rb:208]: If the `CMS::Section` is not `root?`, shouldn't there always be a `parent`? Maybe we can remove the safe navigator.

**Review by thalesmiguel — COMMENTED** [review]
Just a couple comments.

File: app/services/admin/api/cms/template_service.rb (hunk 5 above)
Comment (jlledom) [reply app/services/admin/api/cms/template_service.rb]: Done

File: app/controllers/provider/admin/cms/sections_controller.rb (hunk 7 above)
Comment (jlledom) [reply app/controllers/provider/admin/cms/sections_controller.rb:67]: The bug was also present in API endpoints for both Files and Sections. I solved it by adding a validation to the models. I thought it was a better place than controllers because it's in fact a validation problem and I don't think it's good to repeat the same logic (manually filtering between current provider sections) in every controller. This is a summary of what I found:

- Sections:
  - In both API and UI it was accepting any value, even non-existing section IDs
  - It doesn't fail in master, from the API (where the code to manually filter parents by provider is not present), because the `parent_id` parameter is not in the `attr_accessible` list and thus always ignored.
- Files:
  - In both API and UI it was accepting other providers sections, but not non-existing sections. Trying to set an non-existing section returned an error about section being blank.
  - It doesn't fail in master, from the API, due to some Rails black magic that makes every non-existing or existing-but-not-same-provider sections return a "Section can't be blank" error. For some reason, removing `attr_accessible` for the model caused sections belonging to other providers to be accepted, but non-existing sections still return the "Blank section" error. ¯\\_(ツ)_/¯

For templates there's no problem because `TemplateService` contains the logic to filter the given section.

File: app/controllers/admin/api/cms/templates_controller.rb (hunk 10 above)
Comment (jlledom) [reply app/controllers/admin/api/cms/templates_controller.rb:5]: Yes, this list is used to filter permitted params for mass-assignment, which must have the same name as DB columns. `layout_name` and `section_name` are still accepted but are not permitted for mass-assignment, `TemplateService` takes care of them manually.

File: app/controllers/admin/api/cms/templates_controller.rb [hunk 14]
```diff
@@ -2,12 +2,19 @@ class Admin::Api::CMS::TemplatesController < Admin::Api::CMS::BaseController
...
+    'builtin-partial': %i[draft],
+  }.freeze
 
-  wrap_parameters :template, include: ALLOWED_PARAMS,
-                             format: [:json, :xml, :multipart_form, :url_encoded_form]
+  wrap_parameters :template, include: AVAILABLE_PARAMS,
+                             format: %i[json xml multipart_form url_encoded_form]
 
-  before_action :find_template, :except => [ :index, :create ]
+  before_action :find_template, :except => %i[index create]
```
[the remark this answers is not in the document]
Comment (jlledom) [reply app/controllers/admin/api/cms/templates_controller.rb]: Commited!

File: app/controllers/admin/api/cms/files_controller.rb (hunk 11 above)
Comment (jlledom) [reply app/controllers/admin/api/cms/files_controller.rb]: The tags feature is not supported in the API endpoints. I created an issue ([THREESCALE-8996](https://issues.redhat.com/browse/THREESCALE-8996)) to add it but after a debate in slack we decided to not do it. I removed the tags from the file endpoints for consistency.

File: app/models/cms/section.rb (hunk 13 above)
Comment (jlledom) [reply app/models/cms/section.rb:208]: If it remove the safe operator, this test fails: [test/unit/cms/section_test.rb#L11-L15](https://github.com/3scale/porta/blob/master/test/unit/cms/section_test.rb#L11-L15)

File: app/controllers/admin/api/cms/files_controller.rb (hunk 12 above)
Comment (jlledom) [reply app/controllers/admin/api/cms/files_controller.rb:18]: Exactly

File: app/models/cms/section.rb (hunk 13 above)
Comment (thalesmiguel) [reply app/models/cms/section.rb:208]: Ah OK... It fails because the test object is a new record and `parent_id` is null. For that case, [this validation](https://github.com/3scale/porta/blob/master/app/models/cms/section.rb#L24) will fire and we'll get a `parent_id can't be blank`.

File: app/controllers/admin/api/cms/templates_controller.rb [hunk 15]
```diff
@@ -132,11 +139,21 @@ def find_template
...
   private
 
+  def allowed_type_params
+    ALLOWED_PARAMS[template_type]
+  end
+
+  def template_type
+    return params[:type].parameterize.to_sym if params[:type].present?
+
+    @template.class.name[5..-1].parameterize.to_sym if @template.present?
```
Comment (thalesmiguel) [inline app/controllers/admin/api/cms/templates_controller.rb:154]: Wouldn't this be the same as
```suggestion
    @template.class.data_tag if @template.present?
```
?

File: app/controllers/admin/api/cms/templates_controller.rb [hunk 16]
```diff
@@ -132,11 +139,21 @@ def find_template
...
   end
 
   private
 
+  def allowed_type_params
+    ALLOWED_PARAMS[template_type]
+  end
+
+  def template_type
+    return params[:type].parameterize.to_sym if params[:type].present?
```
Comment (thalesmiguel) [inline app/controllers/admin/api/cms/templates_controller.rb:152]: Just to double check: If an unknown type is provided, `Admin::Api::CMS::TemplateService` will be responsible for raising the error, right? Also, is the error displayed in the API response? I mean, do we have some some rescue to return the error and a 4xx status?

**Review by thalesmiguel — COMMENTED** [review]
I think I'm done with the review. Just asked a couple more questions but this looks good to me 👍

File: app/controllers/admin/api/cms/templates_controller.rb (hunk 15 above)
Comment (jlledom) [reply app/controllers/admin/api/cms/templates_controller.rb:154]: It's not exactly the same, for instance, for builtin partials this returns `:'builtin-partial'` while `data_tag` is `:builtin_partial`. I could just use the data tags and change the keys in the `ALLOWED_PARAMS` hash. But I don't think that's correct because then we are using the data tags for something they were not intended for. Data tags are to show in responses, let's use them only for that.

File: app/controllers/admin/api/cms/templates_controller.rb (hunk 16 above)
Comment (jlledom) [reply app/controllers/admin/api/cms/templates_controller.rb:152]: yes, an error will be raised from `TemplateService` and captured at [app/controllers/admin/api/cms/templates_controller.rb#L61](https://github.com/3scale/porta/blob/9f628f67d67cbad5b364d4213e44cb1c582076ee/app/controllers/admin/api/cms/templates_controller.rb#L61)

File: app/controllers/admin/api/cms/templates_controller.rb (hunk 15 above)
Comment (thalesmiguel) [reply app/controllers/admin/api/cms/templates_controller.rb:154]: > because then we are using the data tags for something they were not intended for. Data tags are to show in responses, let's use them only for that.

Agreed! Then it shouldn't be used here **for sure**.

I was wondering if it'd make sense to use `#type` as keys in here but I don't think it's available for all `CMS::` classes. The motivation is to remove this exclusive logic from the controller in favour of something that makes more sense for the `CMS::` fellas.
Anyways, maybe it doesn't make sense to add yet more code there for the models.

Comment (jlledom) [reply app/controllers/admin/api/cms/templates_controller.rb:154]: The `type` is in fact already being used as keys in the hash, it only computes the key as `@template.class.name[5..-1].parameterize.to_sym` when the `type` param isn't received.
